src/bigistore.c

```diff
@@ -647,6 +648,185 @@ Datum bigistore_each(PG_FUNCTION_ARGS)
+    Assert(rsinfo->returnMode == SFRM_Materialize);
+
+    oldcontext          = MemoryContextSwitchTo(rsinfo->econtext->ecxt_per_query_memory);
+    tupstore            = tuplestore_begin_heap(true, false, work_mem);
+    rsinfo->setResult   = tupstore;
+    rsinfo->setDesc     = tupdesc;
+    MemoryContextSwitchTo(oldcontext);
+
+    /* deconstruct argument array */
+    input       = PG_GETARG_ARRAYTYPE_P(0);
+    i_eltype    = ARR_ELEMTYPE(input);
+    get_typlenbyvalalign(i_eltype, &i_typlen, &i_typbyval, &i_typalign);
+    deconstruct_array(input, i_eltype, i_typlen, i_typbyval, i_typalign,
+                      &i_data, &i_nulls, &arg_count );
+
+    offset  = palloc0(sizeof(int) * arg_count);
+    len     = palloc0(sizeof(int) * arg_count);
+    pairs   = palloc0(sizeof(BigIStorePair *) * arg_count);
+    dvalues = palloc0(sizeof(Datum) * (arg_count + 1));
+    nulls   = palloc0(sizeof(bool) * (arg_count + 1));
+
+    for(int i = 0; i < arg_count; i++ )
+    {
+        BigIStore *bigistore = (BigIStore *) i_data[i];
+
+        len[i]      = bigistore->len;
+        pairs[i]    = FIRST_PAIR(bigistore, BigIStorePair);
+    }
+
+    for(;;)
+    {
+        int32 current_key;
+
+        memset(nulls, false, (arg_count+1) * sizeof(bool));
+        /* set first key to null to indicate end */
+        nulls[0] = true;
+
+        for(int i = 0; i < arg_count; i++ )
+        {
+            if(offset[i] >= len[i])
+            {
+                nulls[i+1] = true;
+                continue;
+            }
+            else
+            {
+                if(nulls[0])
+                {
+                    nulls[0] = false;
+                    current_key = pairs[i]->key;
+                }
+            }
+
+            /* when key matches take it */
+            if(current_key == pairs[i]->key)
+            {
+                dvalues[i+1] = Int64GetDatum(pairs[i]->val);
+                offset[i]++;
+                pairs[i]++;
+            }
+            /* if current_key is to big, all already set non null values must be reset */
+            else if(current_key > pairs[i]->key)
+            {
+                current_key = pairs[i]->key;
+                dvalues[i+1] = Int64GetDatum(pairs[i]->val);
+                offset[i]++;
+                pairs[i]++;
+
+                /* set all we have seen so far to NULL*/
+                for(int j = 0; j < i; j++)
+                {
+                    if(!nulls[j+1])
+                    {
+                        offset[j]--;
+                        pairs[j]--;
+                        nulls[j+1] = true;
+                    }
+                }
+            }
+            else /* just wait for the next round this one is null*/
+            {
+                nulls[i+1] = true;
+            }
+
+        }
+        if (nulls[0])
+            break;
+
+        dvalues[0] = Int32GetDatum(current_key);
+        tuple      = heap_form_tuple(tupdesc, dvalues, nulls);
```

Comment:
Hm... so we just toss a row of N + 1 64bit integers and hope PostgreSQL will fit it into whatever type description the caller provides. That doesn't produce meaningful results unless the caller somehow guessed it needs to provide a description of N+1 integers, and actually crashes when the caller is creative and provides, say, an array or other varlena  as one of the output types:
```
alexk@[local]/postgres # select * from join('1=>1, 2=>1', '1=>3', '3=>4') as t(a text);
server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
The connection to the server was lost. Attempting reset: Failed.
```
Instead of doing this, we should declare the function in SQL as `returns table (key bigint, values bigint[])` and rely on this structure inside the function, calling construct_array to produce values argument.

Reply:
agree that a crash is not nice but what we want is a result of n+1 columns 
maybe we can create separate function for the common case joining upto 10 istores and leave
the array option for if one need to join more than 10 istores.
Or can we check that that the type description makes sense and bail out with a descriptive error if not?

Reply:
Yeah, we can check the type description to have N + 1 bigint columns and emit an error otherwise.

Reply:
Also, the question is why do you need a multi-column result? it's also trivial to convert an array representation to the additional columns by selecting individual elements of the array, i.e select key, values[0], values[1] from (... .);

Reply:
mmmh that's actually a good argument, let's do that then for simplicity